pkg/lib/telemetry/telemetry.go

```diff
@@ -17,121 +17,175 @@ limitations under the License.
+		DefaultContext: &analytics.Context{
+			App: analytics.AppInfo{
+				Version: consts.CortexVersion,
+			},
+			Device: analytics.DeviceInfo{
+				Type: telemetryConfig.Environment,
+			},
+			Extra: map[string]interface{}{
+				"direct": true,
+			},
+		},
+	})
+	if err != nil {
+		_config = nil
+		return err
 	}
-}
 
-type ErrorEvent struct {
-	Timestamp  time.Time `json:"timestamp"`
-	Version    string    `json:"version"`
-	OperatorID string    `json:"operator_id"`
-	Title      string    `json:"title"`
-	Stacktrace string    `json:"stacktrace"`
+	_config = &telemetryConfig
+	return nil
 }
 
-func (c *Client) sendErrorEvent(err error) {
-	errorEvent := ErrorEvent{
-		Timestamp:  time.Now(),
-		Version:    consts.CortexVersion,
-		OperatorID: c.OperatorID,
-		Title:      err.Error(),
-		Stacktrace: fmt.Sprintf("%+v", err),
+func ReportEvent(name string, properties map[string]interface{}) {
+	if _config == nil {
+		return
 	}
-	byteArray, _ := json.Marshal(errorEvent)
-	resp, err := c.http.Post(c.url+errorPath, "application/json", bytes.NewReader(byteArray))
+
+	err := _segment.Enqueue(analytics.Track{
+		Event:      name,
+		UserId:     _config.UserID,
+		Properties: properties,
+	})
 	if err != nil {
-		errors.PrintError(err)
+		ReportError(err)
+	}
+}
+
+func ReportError(err error) {
+	if err == nil || _config == nil {
+		return
+	}
+
+	sentry.WithScope(func(scope *sentry.Scope) {
+		scope.SetUser(sentry.User{ID: _config.UserID})
+		sentry.CaptureException(err)
+		go sentry.Flush(10 * time.Second)
+	})
+}
+
+func RecordEmail(email string) {
+	if _config == nil {
 		return
 	}
 
-	defer resp.Body.Close()
+	_segment.Enqueue(analytics.Identify{
+		UserId: _config.UserID,
+		Traits: analytics.NewTraits().
+			SetEmail(email),
+	})
+}
 
-	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		if byteArray, err := ioutil.ReadAll(resp.Body); err == nil {
-			fmt.Println(string(byteArray))
-		}
+func RecordOperatorID(clientID string, operatorID string) {
+	if _config == nil {
+		return
 	}
+
+	_segment.Enqueue(analytics.Identify{
+		UserId: clientID,
+		Traits: analytics.NewTraits().
+			Set("operator_id", operatorID),
+	})
 }
 
-func (c *Client) ReportEvent(name string) {
-	if c.EnableTelemetry {
-		go c.sendUsageEvent(name)
+func closeSentry() error {
+	if !sentry.Flush(5 * time.Second) {
+		return errors.New("sentry flush timout exceeded")
 	}
+	return nil
 }
 
-func (c *Client) ReportErrorBlocking(err error) {
-	if c.EnableTelemetry {
-		c.sendErrorEvent(err)
+func closeSegment() error {
+	if _segment == nil {
+		return nil
 	}
+	return _segment.Close()
+}
+
+func Close() {
+	parallel.Run(closeSegment, closeSentry)
+	_config = nil
 }
 
-func (c *Client) ReportError(err error) {
-	if c.EnableTelemetry {
-		go c.sendErrorEvent(err)
+func ExitErr(errs ...interface{}) {
+	err := errors.MergeErrItems(errs...)
+	if err == nil {
+		err = errors.New("blank")
 	}
+	ReportError(err)
+
+	Close()
+
+	errors.Exit(errs...)
+}
+
+func ExitOk() {
```

Comment:
Maybe this function should be in root.go or another file. I don't think exiting code should live in telemetry.